spyder/plugins/completion/languageserver/providers/utils.py

```diff
@@ -6,22 +6,53 @@
 
 """Spyder Language Server Protocol Client common util functions."""
 
-
+import os
 import os.path as osp
 from spyder.py3compat import PY2
+from spyder.utils.encoding import to_unicode
+
 
 if PY2:
     import pathlib2 as pathlib
     from urlparse import urlparse
     from urllib import url2pathname
+    from urllib import quote as urlquote_from_bytes
 else:
     import pathlib
     from urllib.parse import urlparse
+    from urllib.parse import quote_from_bytes as urlquote_from_bytes
     from urllib.request import url2pathname
 
 
+def as_posix(path_obj):
+    """Get path as_posix as unicode using correct separator."""
+    path = path_obj
+    return to_unicode(str(path)).replace(os.sep, '/')
+
+
+def make_as_uri(path):
+    """
+    As URI path for Windows.
+    Reimplementation of Path.make_as_uri to use an unicode as_posix version.
```

Comment:
Could you elaborate more on what this does?

Reply:
Basically one of the encoding errors was caused because of the way pathlib2 creates the path representation in the `as_posix` method and then tries to encode it in the `make_as_uri` method:
```python
    def make_uri(self, path):
        # Under Windows, file URIs use the UTF-8 encoding.
        drive = path.drive
        if len(drive) == 2 and drive[1] == ':':
            # It's a path on a local drive => 'file:///c:/a/b'
            rest = path.as_posix()[2:].lstrip('/')
            return 'file:///%s/%s' % (
                drive, urlquote_from_bytes(rest.encode('utf-8')))
        else:
            # It's a path on a network drive => 'file://host/share/a/b'
            return 'file:' + urlquote_from_bytes(
                path.as_posix().encode('utf-8'))
```

```python
def as_posix(self):
        """Return the string representation of the path with forward (/)
        slashes."""
        f = self._flavour
        return str(self).replace(f.sep, '/')
```

The `make_as_uri` method (that triggered an UnicodeDecodeError) is called when we were calling the `as_uri` method of a Path Object at `path_as_uri`. On Windows the call chain is something like `utils.path_as_uri()` > `Path.as_uri()` > `WindowsPath.make_as_uri()` > UnicodeDecodeError

This is the traceback you would get:
```traceback
  File "C:\Users\Daniel\Documents\Spyder\spyder\spyder\plugins\editor\plugin.py", line 283, in report_open_file
    language.lower(), filename, codeeditor)
  File "C:\Users\Daniel\Documents\Spyder\spyder\spyder\plugins\completion\plugin.py", line 245, in register_file
    language, filename, codeeditor
  File "C:\Users\Daniel\Documents\Spyder\spyder\spyder\plugins\completion\languageserver\plugin.py", line 63, in register_file
    language_client.register_file(filename, codeeditor)
  File "C:\Users\Daniel\Documents\Spyder\spyder\spyder\plugins\completion\languageserver\providers\document.py", line 29, in register_file
    filename = path_as_uri(filename)
  File "C:\Users\Daniel\Documents\Spyder\spyder\spyder\plugins\completion\languageserver\providers\utils.py", line 55, in path_as_uri
    return path_obj.as_uri()
  File "C:\Users\Daniel\Anaconda3\envs\spyder27\lib\site-packages\pathlib2\__init__.py", line 970, in as_uri
    return self._flavour.make_uri(self)
  File "C:\Users\Daniel\Anaconda3\envs\spyder27\lib\site-packages\pathlib2\__init__.py", line 430, in make_uri
    drive, urlquote_from_bytes(rest.encode('utf-8')))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xed in position 18: ordinal not in range(128)

```